#**Hey Kagglers, this notebook will go over a couple of basic Natural Language Processing techniques using the [scikit-learn](http://scikit-learn.org/stable/) library. Special thanks to [Aaron7sun](https://www.kaggle.com/aaron7sun) for providing the [dataset](https://www.kaggle.com/aaron7sun/stocknews) and "assignment" for us to do!**  
#Thanks for reading and feel free to leave feedback. I'd like to make more tutorial notebooks like this, so let me know if you think there is anything I could improve.

----------

# Notebook Prep

First things first, let's import the libraries we'll be using.  

In [1]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression

[Pandas](http://pandas.pydata.org/) will make our data easy to look at and work with.  
[CountVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html), part of scikit-learn, will take care of our NLP tasks.  
[LogisticRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html), also part of scikit-learn, will train and test our predictive models.

----------

# Data Import

Now, let's [read](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) in the data with Pandas.  
If you're working in something other than a Kaggle notebook, be sure to change the file location.  
For this tutorial, we're just going to use the combined dataset that Aaron prepared for us, but you're welcome to import the other two CSV files if you want to combine them in a different way.

In [6]:
data = pd.read_csv('../input/Combined_News_DJIA.csv',parse_dates=[0])

Next, let's take a look at the data with the [head](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html) method.

In [10]:
data['Date'].max()

Timestamp('2016-07-01 00:00:00')

We've got a lot of vaiables here, but the layout is pretty straight-forward.  
As a reminder, the Label variable will be a **1** if the DJIA **stayed the same or rose** on that date or 
 **0** if the DJIA **fell** on that date.

And finally, before we get started on the rest of the notebook, we need to split our data into a training set and a testing set. Per Aaron's instructions, we'll use all of the dates up to the end of 2014 as our training data and everything after as testing data.

In [11]:
train = data[data['Date'] < '2015-01-01']
test = data[data['Date'] > '2014-12-31']

----------

# Text Preprocessing

Now that our data is loaded in, we need to clean it up just a little bit to prepare it for the rest of our analysis.  
To illustrate this process, look at how the example headline below changes from cell to cell.  
Don't worry about the code too much here, since this example is only meant to be visual.

In [12]:
example = train.iloc[3,10]
print(example)

b"The commander of a Navy air reconnaissance squadron that provides the President and the defense secretary the airborne ability to command the nation's nuclear weapons has been relieved of duty"


In [13]:
example2 = example.lower()
print(example2)

b"the commander of a navy air reconnaissance squadron that provides the president and the defense secretary the airborne ability to command the nation's nuclear weapons has been relieved of duty"


In [14]:
example3 = CountVectorizer().build_tokenizer()(example2)
print(example3)

['the', 'commander', 'of', 'navy', 'air', 'reconnaissance', 'squadron', 'that', 'provides', 'the', 'president', 'and', 'the', 'defense', 'secretary', 'the', 'airborne', 'ability', 'to', 'command', 'the', 'nation', 'nuclear', 'weapons', 'has', 'been', 'relieved', 'of', 'duty']


In [15]:
pd.DataFrame([[x,example3.count(x)] for x in set(example3)], columns = ['Word', 'Count'])

Word  Count
0         squadron      1
1         provides      1
2             duty      1
3        commander      1
4              the      5
5          weapons      1
6              air      1
7               to      1
8             navy      1
9        secretary      1
10         command      1
11         defense      1
12        airborne      1
13         ability      1
14       president      1
15  reconnaissance      1
16         nuclear      1
17             and      1
18        relieved      1
19          nation      1
20              of      2
21            been      1
22            that      1
23             has      1

Were you able to see everything that changed?  
The process involved:  
- Converting the headline to lowercase letters  
- Splitting the sentence into a list of words  
- Removing punctuation and meaningless words  
- Transforming that list into a table of counts

What started as a relatively "messy" sentence has now become an neatly organized table!  
And while this may not be exactly what goes on behind the scenes with scikit-learn, this example should give you a pretty good idea about how it works.

So now that you've seen what the text processing looks like, let's get started on the fun part, modeling!

----------

# Basic Model Training and Testing

As mentioned previously, scikit-learn is going to take care of all of our preprocessing needs.  
The tool we'll be using is CountVectorizer, which takes a single list of strings as input, and produces word counts for each one.

You might be wondering if our dataframe meets this "single list of strings" criteria, and the answer to that is... it doesn't!  
In order to meet this criteria, we'll use the following [for loop](https://wiki.python.org/moin/ForLoop) to iterate through each row of our dataset, [combine](https://docs.python.org/3.5/library/stdtypes.html#str.join) all of our headlines into a single string, then [add](https://docs.python.org/3.5/tutorial/datastructures.html) that string to the list we need for CountVectorizer.

In [22]:
trainheadlines = []
for row in range(0,len(train.index)):
    trainheadlines.append(' '.join(str(x) for x in train.iloc[row,2:27]))

In [21]:
trainheadlines[:5]

['b"Georgia \'downs two Russian warplanes\' as countries move to brink of war" b\'BREAKING: Musharraf to be impeached.\' b\'Russia Today: Columns of troops roll into South Ossetia; footage from fighting (YouTube)\' b\'Russian tanks are moving towards the capital of South Ossetia, which has reportedly been completely destroyed by Georgian artillery fire\' b"Afghan children raped with \'impunity,\' U.N. official says - this is sick, a three year old was raped and they do nothing" b\'150 Russian tanks have entered South Ossetia whilst Georgia shoots down two Russian jets.\' b"Breaking: Georgia invades South Ossetia, Russia warned it would intervene on SO\'s side" b"The \'enemy combatent\' trials are nothing but a sham: Salim Haman has been sentenced to 5 1/2 years, but will be kept longer anyway just because they feel like it." b\'Georgian troops retreat from S. Osettain capital, presumably leaving several hundred people killed. [VIDEO]\' b\'Did the U.S. Prep Georgia for War with Russia?\

With our headlines formatted, we can set up our CountVectorizer.  
To start, let's just use the default settings and see how it goes!  
Below, we'll name our default vectorizer, then [use](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html#sklearn.feature_extraction.text.CountVectorizer.fit_transform) it on our list of combined headlines.  
After that, we'll take a look at the size of the result to see how many words we have.

In [17]:
basicvectorizer = CountVectorizer()
basictrain = basicvectorizer.fit_transform(trainheadlines)
print(basictrain.shape)

(1611, 31675)


In [20]:
basictrain

<1611x31675 sparse matrix of type '<class 'numpy.int64'>'
	with 489193 stored elements in Compressed Sparse Row format>

Wow! Our resulting table contains counts for 31,675 different words!

Now, let's train a logistic regression model using this data.  
In the cell below, we're simply naming our model, then [fitting](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.fit) the model based on our X and Y values.

In [18]:
basicmodel = LogisticRegression()
basicmodel = basicmodel.fit(basictrain, train["Label"])

/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Our model is ready to go, so let's set up our test data.  
Here, we're just going to repeat the steps we used to prep our training data, then [predict](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict) whether the DJIA increased or decreased for each day in the test dataset.

In [23]:
testheadlines = []
for row in range(0,len(test.index)):
    testheadlines.append(' '.join(str(x) for x in test.iloc[row,2:27]))
basictest = basicvectorizer.transform(testheadlines)
predictions = basicmodel.predict(basictest)

The predictions are set, so let's use a [crosstab](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.crosstab.html) to take a look at the results!

Prediction accuracy is just over 42%. It seems like this model isn't too reliable.  
Now, let's also take a look at the coefficients of our model. (Excellent request from [Lucie](https://www.kaggle.com/luciegattepaille)!)

In [24]:
pd.crosstab(test["Label"], predictions, rownames=["Actual"], colnames=["Predicted"])

Predicted   0    1
Actual            
0          61  125
1          92  100

The cell below will get a list of the names from our CountVectorizer and a list of the coefficients from our model, then combine the two lists into a Pandas dataframe.  
Once that's made, we can sort it and check out the top 10 positive and negative coefficients.

In [25]:
basicwords = basicvectorizer.get_feature_names()
basiccoeffs = basicmodel.coef_.tolist()[0]
coeffdf = pd.DataFrame({'Word' : basicwords, 
                        'Coefficient' : basiccoeffs})
coeffdf = coeffdf.sort_values(['Coefficient', 'Word'], ascending=[0, 1])
coeffdf.head(10)

Word  Coefficient
19419    nigeria     0.497872
25261       self     0.452523
29286         tv     0.427959
15998      korea     0.425862
20135   olympics     0.425712
15843      kills     0.411638
26323         so     0.411301
29256       turn     0.394834
10874      fears     0.388540
28274  territory     0.384034

In [26]:
coeffdf.tail(10)

Word  Coefficient
27299   students    -0.424441
8478         did    -0.427078
6683       congo    -0.431908
12818    hacking    -0.444041
7139     country    -0.448565
16949        low    -0.463092
3651       begin    -0.470475
25433        sex    -0.494526
24754  sanctions    -0.549718
24542        run    -0.587751

Our most positive words don't seem particularly interesting, however there are some negative sounding words within our bottom 10, such as "sanctions," "low," and "hacking."  
Maybe the saying "no news is good news" is true here?

----------

# Advanced Modeling

The technique we just used is known as a **bag-of-words** model. We essentially placed all of our headlines into a "bag" and counted the words as we pulled them out.  
However, most people would agree that a single word doesn't always have enough meaning by itself.  
Obviously, we need to consider the rest of the words in the sentence as well!  

This is where the **n-gram** model comes in.  
In this model, n represents the length of a sequence of words to be counted.  
This means our bag-of-words model was the same as an n-gram model where n = 1.  
So now, let's see what happens when we run an n-gram model where n = 2.

Below, we'll create a new CountVectorizer with the n-gram parameter set to 2 instead of the default value of 1.

In [27]:
advancedvectorizer = CountVectorizer(ngram_range=(2,2))
advancedtrain = advancedvectorizer.fit_transform(trainheadlines)

Now that we've run our vectorizer, let's see what our data looks like this time around.

In [28]:
print(advancedtrain.shape)

(1611, 366721)


This time we have 366,721 unique variables representing two-word combinations!  
And here I thought last time was big...

So, just like last time, let's name and fit our regression model.

In [29]:
advancedmodel = LogisticRegression()
advancedmodel = advancedmodel.fit(advancedtrain, train["Label"])

/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


And again like last time, let's transform our test data and make some predictions!

In [30]:
testheadlines = []
for row in range(0,len(test.index)):
    testheadlines.append(' '.join(str(x) for x in test.iloc[row,2:27]))
advancedtest = advancedvectorizer.transform(testheadlines)
advpredictions = advancedmodel.predict(advancedtest)

Crosstab says...!

In [31]:
pd.crosstab(test["Label"], advpredictions, rownames=["Actual"], colnames=["Predicted"])

Predicted   0    1
Actual            
0          66  120
1          45  147

This time we're up to nearly 57% prediction accuracy.  
We might only consider this a slight improvement, but keep in mind that we've barely scratched the surface of NLP here, and we haven't even touched more advanced machine learning techniques.  
Let's check out our coefficients again as well!

In [32]:
advwords = advancedvectorizer.get_feature_names()
advcoeffs = advancedmodel.coef_.tolist()[0]
advcoeffdf = pd.DataFrame({'Words' : advwords, 
                        'Coefficient' : advcoeffs})
advcoeffdf = advcoeffdf.sort_values(['Coefficient', 'Words'], ascending=[0, 1])
advcoeffdf.head(10)

Words  Coefficient
272047   right to     0.286533
24710   and other     0.275274
285392     set to     0.274698
316194  the first     0.262873
157511   in china     0.227943
159522   in south     0.224184
125870   found in     0.219130
124411  forced to     0.216726
173246     it has     0.211137
322590    this is     0.209239

In [33]:
advcoeffdf.tail(10)

Words  Coefficient
326846      to help    -0.198495
118707      fire on    -0.201654
155038        if he    -0.209702
242528   people are    -0.211303
31669    around the    -0.213362
321333     there is    -0.215699
327113      to kill    -0.221812
340714        up in    -0.226289
358917    with iran    -0.227516
315485  the country    -0.331153

It seems that the results this time were fairly similar. Most of the positive bigrams are unremarkable, while a few of the negative ones like "bin laden" and "threatens to" could be considered to carry some negative meaning.

----------

# What's next?

If you'd like to keep going forward with this notebook, here are a couple of project ideas:  
- Experiment with different n values using the n-gram model  
- Use previous days' headlines to truly "predict" whether the DJIA will rise or fall  
- Try a machine learning algorithm instead of the basic logistic regression used in this notebook

Thanks again for reading! I hope you found this helpful!